### PR TITLE
Amend transactions v2 proposal with flag to disable signer propagation

### PR DESCRIPTION
**Pull request by jstarry**
#### Problem
As outlined in https://github.com/solana-labs/solana/issues/17762, the default runtime behavior of allowing signers to be propagated to inner instructions can be taken advantage of by malicious parties. For example, if a user passes the transaction signer account to a malicious program, that program can invoke the system program to drain the signer's funds.

#### Summary of Changes
- Proposed adding wallet-configurable flags to each transaction instruction. The first (and only) flag is the `DISABLE_SIGNER_PROPAGATION` flag which will not allow invoked programs to propagate the signer privileges that they received from their caller when invoking other programs.

Addresses: https://github.com/solana-labs/solana/issues/15323
Addresses: https://github.com/solana-labs/solana/issues/17762

### CODE REVIEW

File: docs/src/proposals/transactions-v2.md [hunk 1]
```diff
@@ -158,15 +157,73 @@ pub struct AddressMap {
...
+pub struct CompiledInstructionV2 {
+  /// NEW: Modify runtime behavior for this instruction
+  pub config: IxConfig,
+
+  /// unchanged
+  pub program_id_index: u8,
+
+  /// NEW: Only programs in this list may be invoked by this instruction
+  #[serde(with = "short_vec")]
+  pub invokable_program_id_indexes: Vec<u8>,
```
Comment (jackcmay) [inline docs/src/proposals/transactions-v2.md:184]: This kinda but I guess not sufficiently covers the need for the bit flags

Comment (jackcmay) [reply docs/src/proposals/transactions-v2.md:184]: The bit flag concept allows for a limited number of other flags/uses cases in the future, so helpful in general.  But for signer propagation, another possibility is to discriminately allow signer propagation by ordering the list of invokable ids and a count of how many of those should the signing be propagated to.  Kinda like ro vs rw accounts in the message.

Comment (jstarry) [reply docs/src/proposals/transactions-v2.md:184]: The motivation behind disabling signer propagation is more around ensuring that a program can't use the system or token program to drain my assets. Most times, the system and token programs going to be the programs which are invoked with signer propagation so I don't think that level of granularity that way adds much safety to users.

**Comment by aeyakovenko** [comment]
Will these always be set to full rights though?  Now that phantom shows the pre/post impact, those attack vectors have died down.

**Comment by jstarry** [comment]
> Will these always be set to full rights though? Now that phantom shows the pre/post impact, those attack vectors have died down.

The proposal to add a configurable flag was designed to provide a migration path to allow programs time to change their APIs to not use signer propagation. Wallets can set these params on behalf of users as individual programs become compatible.

**Comment by fragosti** [comment]
Appreciate all attempts at trying to make things safer!

Generally like the direction of allowing wallets to provide more guarantees – I do wish we could make guarantees about maximum amounts to the system and SPL programs instead, since those are easier to reason about (for users as well).

This direction pushes more responsibility onto wallets and dApps to understand the implementation details of the programs. In the case of the whitelist we need to know which things to whitelist. In the case of the flag, we need to know when to enable it (is there even a clear answer here?).

These decisions propagate up to users as well. Instead of being able to say "hey, you're going to transfer MAX 5 SOL here", we have to show a warning that has all these caveats and technicalities. I have personally struggled to come up with copy here. For example, explaining the "dangling approval" problem is hard.

**Comment by jstarry** [comment]
> Appreciate all attempts at trying to make things safer!
> 
> Generally like the direction of allowing wallets to provide more guarantees – I do wish we could make guarantees about maximum amounts to the system and SPL programs instead, since those are easier to reason about (for users as well).
> 
> This direction pushes more responsibility onto wallets and dApps to understand the implementation details of the programs. In the case of the whitelist we need to know which things to whitelist. In the case of the flag, we need to know when to enable it (is there even a clear answer here?).

I think that dApps already have basically full responsibility for knowing how programs will execute a transaction so a bit more responsibility doesn't seem like a big deal there. As for wallets, I don't think that this change adds anymore responsibility because a given client will know whether the transaction it constructed and the programs it invokes can be processed without signer propagation and will set that flag itself. I imagine that after some amount of time for migration, the only responsibility that wallets will have is indicating to users that the transaction they are signing is unsafe if the flag isn't set.

> These decisions propagate up to users as well. Instead of being able to say "hey, you're going to transfer MAX 5 SOL here", we have to show a warning that has all these caveats and technicalities. I have personally struggled to come up with copy here. For example, explaining the "dangling approval" problem is hard.

100% agree here. This is always going to be a cat and mouse game but I'm hoping we can work towards limiting the amount of damage that can be done. I think that dangling approvals are another solvable problem which I'll follow up with a new proposal for soon.

**Comment by fragosti** [comment]
> > Appreciate all attempts at trying to make things safer!
> > Generally like the direction of allowing wallets to provide more guarantees – I do wish we could make guarantees about maximum amounts to the system and SPL programs instead, since those are easier to reason about (for users as well).
> > This direction pushes more responsibility onto wallets and dApps to understand the implementation details of the programs. In the case of the whitelist we need to know which things to whitelist. In the case of the flag, we need to know when to enable it (is there even a clear answer here?).
> 
> I think that dApps already have basically full responsibility for knowing how programs will execute a transaction so a bit more responsibility doesn't seem like a big deal there. As for wallets, I don't think that this change adds anymore responsibility because a given client will know whether the transaction it constructed and the programs it invokes can be processed without signer propagation and will set that flag itself. I imagine that after some amount of time for migration, the only responsibility that wallets will have is indicating to users that the transaction they are signing is unsafe if the flag isn't set.
> 
> > These decisions propagate up to users as well. Instead of being able to say "hey, you're going to transfer MAX 5 SOL here", we have to show a warning that has all these caveats and technicalities. I have personally struggled to come up with copy here. For example, explaining the "dangling approval" problem is hard.
> 
> 100% agree here. This is always going to be a cat and mouse game but I'm hoping we can work towards limiting the amount of damage that can be done. I think that dangling approvals are another solvable problem which I'll follow up with a new proposal for soon.

Thanks! I'm all for incremental solutions fwiw. Figured I'd take a chance to paint a bigger picture. Is the expectation that we'd show a warning if the flag wasn't enabled or something?

**Pull request closed without merging**